buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalBwcGitPlugin.java

```diff
@@ -0,0 +1,202 @@
+
+        tasks.register("checkoutBwcBranch", checkoutBwcBranch -> {
+            checkoutBwcBranch.dependsOn(fetchLatestTaskProvider);
+            checkoutBwcBranch.doLast(t -> {
+                Logger logger = project.getLogger();
+
+                String bwcBranch = this.gitExtension.getBwcBranch().get();
+                final String refspec = providers.systemProperty("bwc.refspec." + bwcBranch)
+                    .orElse(providers.systemProperty("tests.bwc.refspec." + bwcBranch))
+                    .getOrElse(remote.get() + "/" + bwcBranch);
+
+                String effectiveRefSpec = maybeAlignedRefSpec(project, logger, refspec);
+
+                logger.lifecycle("Performing checkout of {}...", refspec);
+                LoggedExec.exec(project, spec -> {
+                    spec.workingDir(gitExtension.getCheckoutDir());
+                    spec.commandLine("git", "checkout", effectiveRefSpec);
+                });
+
+                String checkoutHash = GlobalBuildInfoPlugin.gitInfo(gitExtension.getCheckoutDir().get()).getRevision();
+                logger.lifecycle("Checkout hash for {} is {}", project.getPath(), checkoutHash);
+                writeFile(new File(project.getBuildDir(), "refspec"), checkoutHash);
+            });
+        });
+    }
+
+    public BwcGitExtension getGitExtension() {
+        return gitExtension;
+    }
+
+    /**
+     * We use a time based approach to make the bwc versions built deterministic and compatible with the current hash.
+     * Most of the time we want to test against latest, but when running delayed exhaustive tests or wanting
+     * reproducible builds we want this to be deterministic by using a hash that was the latest when the current
+     * commit was made.
+     * <p>
+     * This approach doesn't work with merge commits as these can introduce commits in the chronological order
+     * after the fact e.x. a merge done today can add commits dated with yesterday so the result will no longer be
+     * deterministic.
+     * <p>
+     * We don't use merge commits, but for additional safety we check that no such commits exist in the time period
+     * we are interested in.
+     * <p>
+     * Timestamps are at seconds resolution. rev-parse --before and --after are inclusive w.r.t the second
+     * passed as input. This means the results might not be deterministic in the current second, but this
+     * should not matter in practice.
+     */
+    private String maybeAlignedRefSpec(Project project, Logger logger, String defaultRefSpec) {
+        if (project.getProviders().systemProperty("bwc.checkout.align").isPresent() == false) {
+            return defaultRefSpec;
+        }
+
+        String timeOfCurrent = execGit(execSpec -> {
+            execSpec.commandLine(asList("git", "show", "--no-patch", "--no-notes", "--pretty='%cD'"));
+            execSpec.workingDir(project.getRootDir());
+        });
+
+        logger.lifecycle("Commit date of current: {}", timeOfCurrent);
```

Comment:
What does the entirety of the lifecycle output look like? This comment on its own would be pretty cryptic

Reply:
The overall lifecycle output here hasn't changed in this PR and would look e.g. like this:

```
> Task :distribution:bwc:bugfix:checkoutBwcBranch
Commit date of current: 'Wed, 30 Sep 2020 10:33:41 +0200'
Performing checkout of elastic/7.9...
Checkout hash for :distribution:bwc:bugfix is c9534e03e8beae7e9b566cda7236532665c376e9
```